src/algebra/char_p/algebra.lean

```diff
@@ -58,6 +58,30 @@ lemma char_zero_of_injective_algebra_map {R A : Type*} [comm_semiring R] [semiri
 -- `char_p.char_p_to_char_zero A _ (char_p_of_injective_algebra_map h 0)` does not work
 -- here as it would require `ring A`.
 
+/-!
+As an application, a `ℚ`-algebra has characteristic zero.
+-/
+section Q_algebra
+
+variables (R : Type*) [nontrivial R]
+
+/-
+Implementation note: These are not instances because the counter direction `algebra_rat` in
+[algebra.algebra.basic] is chosen as instance:
+(`algebra_rat [division_ring α] [char_zero α] : algebra ℚ α`)
+-/
+
+lemma Q_algebra_char_p_zero [semiring R] [algebra ℚ R] : char_p R 0 :=
+char_p_of_injective_algebra_map (algebra_map ℚ R).injective 0
+
+lemma Q_algebra_char_zero [ring R] [algebra ℚ R] : char_zero R :=
+@char_p.char_p_to_char_zero R _ (Q_algebra_char_p_zero R)
```

Comment:
I must say that I can't really think of a case where we already have an instance `algebra ℚ R` on our hands but no `char_zero`. Do you have a specific use case in mind? (I was thinking along the lines of `polynomial ℚ`, but we already directly show that `polynomial` has the same characteristic without going through `algebra`.)

Otherwise these definitions are definitely not suitable as (local) instances as you remark above, so let's add a docstring to warn the users. (I believe using it as `haveI : char_zero R := Q_algebra_char_zero R` is safe, since `char_zero` lives in `Prop` there won't be any defeq issues.)

Another remark is that we assume `semiring R` in the first lemma but `algebra ℚ R` means that `R` contains `-1` so it is at least a ring. This is probably fine, but a little surprising.

A final little thing is that I don't really like the names. How about these?
```suggestion
/-- A nontrivial `ℚ`-algebra has `char_p` equal to zero.

This cannot be a (local) instance because it would immediately form a loop with the instance `algebra_rat`.
It's probably easier to go the other way: prove `char_zero R` and automatically receive an `algebra ℚ R` instance.
-/
lemma algebra_rat.char_p_zero [semiring R] [algebra ℚ R] : char_p R 0 :=
char_p_of_injective_algebra_map (algebra_map ℚ R).injective 0

/-- A nontrivial `ℚ`-algebra has characteristic zero.

This cannot be a (local) instance because it would immediately form a loop with the instance `algebra_rat`.
It's probably easier to go the other way: prove `char_zero R` and automatically receive an `algebra ℚ R` instance.
-/
lemma algebra_rat.char_zero [ring R] [algebra ℚ R] : char_zero R :=
@char_p.char_p_to_char_zero R _ (Q_algebra_char_p_zero R)
```

Reply:
Update: I see, something like this gets used for example in `Q_algebra_to_equal_char_zero`.

Reply:
> Another remark is that we assume semiring R in the first lemma but algebra ℚ R means that R contains -1 so it is at least a ring. This is probably fine, but a little surprising.
Would it be worth adding that comment as something like `instance xyz [semiring R] [algebra ℚ R] : ring R := sorry` in a new PR?

I personally don't care (here in this PR) about semirings and that first lemma, just added it because `char_p _ 0` and `char_zero` run somewhat parallel.